README.md
# SndFile2K

[![Build Status](https://secure.travis-ci.org/evpobr/sndfile2k.svg?branch=master)](http://travis-ci.org/evpobr/sndfile2k)
[![Build status](https://ci.appveyor.com/api/projects/status/7ke6kwrluv5o6ct3/branch/master?svg=true)](https://ci.appveyor.com/project/evpobr/sndfile2k-3tb6i/branch/master)

Fork of [libsndfile](http://www.mega-nerd.com/libsndfile/) library, a C library for reading and writing files containing sampled audio data.

SndFile2K documentation is available [online](https://evpobr.github.io/sndfile2k).

**Warning** this repo is experimental, most likely you want to use [libsndfile](http://www.mega-nerd.com/libsndfile/).

## Getting Started

### Prerequisites
* C99 compatible compiler, GCC and Clang are fine. Visual Studio 2015 or higher supported under Windows.
* [CMake >= 3.2](https://cmake.org/)
* [Ogg](https://xiph.org/ogg/), [Vorbis](http://www.vorbis.com/) and [FLAC](https://xiph.org/flac/) development libraries (optional)
* [Ninja build system](https://ninja-build.org/) (optional)

### Contributing
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

### Versioning
We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/evpobr/sndfile2k/tags/).

## Authors
* [Erik de Castro Lopo](https://github.com/erikd) - *Initial work*
* [evpobr](https://github.com/evpobr) - *Author of this fork*

See also the list of [contributors](https://github.com/evpobr/sndfile2k/contributors) who participated in this project.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2018-03-25

### Fixed

- Skipping of erroneous subchunks
- Always return SF_ERR_UNRECOGNISED_FORMAT on unrecognised format
- Avoid unnessesary seeks when opening Ogg files

### Changed

- Coding style rules: bump column width limit of C source files to 132 chars.
  We have too long lines, so i choose 132 for readability. It is another
  historical limit. More explained [here](https://lkml.org/lkml/2009/12/17/229).
- New Posix based internal file IO. Now IO is unform across all platforms.
- `sf_open_virtual()` function deprecated, use `sf_open_virtual_ex` instead.
- `sf_open_fd()` function deprecated, use `sf_open_virtual_ex` instead.
- Broadcast info support deprecated and will be removed in next release.
- CART info support deprecated and will be removed in next release.

### Added

- New members os `SF_VIRTUAL_IO` struct: `flush`, `set_filelen`, `is_pipe`, `ref`
  and `unref`. Now you can implement fully functional IO with SF_VIRTUAL_IO.
- New `sf_open_virtual_ex()` function to use all the features of `SF_VIRTUAL_IO`.

## [1.1.1] - 2018-03-10

### Added

- NMS ADPCM codec in WAV support

### Fixed

- Max channel count bug

## [1.1.0] - 2018-03-02

### Added

- NMS ADPCM codec support

### Fixed

- `sndfile-convert` fail due to no. of channels
- `SFC_GET_LIB_VERSION` command behavior on error, now it returns 0

### Removed

- MacOSX & BeOS support from `sndfile-play`

## [1.0.0] - 2017-12-24

### Changed

- Coding style rules: bump column width limit of C source files to 100 chars

### Fixed

- Small CMake project fix

## [1.0.0-rc.2] - 2017-12-24

### Added

- CMake documentation target `docs`

### Changed

- CMake installation ia now component based

### Fixed

- Appveyor CI badge in `README.md`
- Documentation glitches

## [1.0.0-rc.1] - 2017-12-22

### Added

- Doxygen generated online documentation available at https://evpobr.github.io/sndfile2k
- CMake `ENABLE_PKGCONFIG_FILE` option

### Changed

- CMake option `DISABLE_CPU_CLIP` name to `ENABLE_CPU_CLIP` and set to `ON` by default
- CMake option `ENABLE_PACKAGE_CONFIG` name to `ENABLE_CMAKE_PACKAGE_CONFIG`
- CMake option `DISABLE_EXTERNAL_LIBS` name to `HAVE_XIPH_CODECS`
- CMake features names
- `sndfile2k.hh` header renamed to `sndfile2k.hpp`
- CMake Ogg, Vorbis and FLAC find modules improved
- `SNDFILE` is now typedef of existing `sf_private_tag` struct

### Fixed

- CMake configuration when `ENABLE_EXPERIMENTAL` is activated but Speex library is not found
- `SFC_GET_LOG_INFO` command return output. When data is NULL, 0 is returned now.
- Some compiler warnings

### Removed

- [Vcpkg](https://github.com/Microsoft/vcpkg) port (issue #7)
- `sndfile-regtest` program removed with SQLite3 dependency

## [1.0.0-beta4] - 2017-11-27

### Added

- [Vcpkg](https://github.com/Microsoft/vcpkg) port in `packages/vcpkg` directory (issue #7)

### Changed

- Improve `pkg-config` module
- Small fixes

## [1.0.0-beta3] - 2017-11-19

### Fixed
- CMake: fix public include directory (issue #6)

### Changed
- Typedef structs and enums
- Small fixes

## [1.0.0-beta2] - 2017-11-08

### Fixed
- Appveyor CI badge
- EditorConfig rules
- Small fixes
- sndfile2k-info linking to winmm library

### Changed
- Unicode open function `sf_wchar_open` is always accessible under Win32 and Cygwin
- Travis CI tests are faster now (OSX test are disabled)

## [1.0.0-beta1] - 2017-11-07

### Fixed

- Compiler errors
- [CMake warning](https://cmake.org/cmake/help/latest/policy/CMP0063.html)

## [1.0.0-alpha1] - 2017-11-07 [YANKED]

### Added

- Tests sources which were generated are now in source tree. No `autogen` stuff requred, faster configuration and compilation.

### Changed

- Version numbering restarted from `1`.
- Project renamed to SndFile2K. Library name is `sndfile2k`. Utilities are renamed in the same way.
- Main public header renamed to sndfile2k to avoid conflicts with installed `libsndfile`.
- Public headers are moved to `include/sndfile2k` directory. It is bad to keep public headers in source directory.
- Public API is now exported via [compiler visibility pragmas](https://gcc.gnu.org/wiki/Visibility), not symbol files. This was done to be maximum portable, also Python is not requred now to build shared library. Since it is different way to export public function, **ABI is not compatible with libsndfile**. You cannot just replace `libsndfile` binary with `sndfile2k` library. Symbols files allows features unsupported via current export system. But **API is compatibile** with `libsndfile`. Recompile and use. Just include `sndfile2k/sndfile2k.h` instead of `sndfile.h`.

  NOTE: You need to add `SNDFILE2K_STATIC` define if you adding **static** SndFile2K library to your project under Windows.
  
- Using reStructuredTest for documentation.
  
### Removed

- Autootools build system in favour of CMake build system.
- Documentation - will be rewritten in ReStructuredText and added later.
- Python based style checking script and git hook - coding style was changed.
- Other unportable or obsolete scripts.
- `Win32` directory. Test file inside was useless, since Windows is supported now.
- Travis CI MacOS GCC test removed

[Unreleased]: https://github.com/evpobr/sndfile2k/compare/v1.2.0...HEAD
[1.0.0-beta1]: https://github.com/evpobr/sndfile2k/compare/v1.0.0-alpha1...v1.0.0-beta1
[1.0.0-beta2]: https://github.com/evpobr/sndfile2k/compare/v1.0.0-beta1...v1.0.0-beta2
[1.0.0-beta3]: https://github.com/evpobr/sndfile2k/compare/v1.0.0-beta2...v1.0.0-beta3
[1.0.0-beta4]: https://github.com/evpobr/sndfile2k/compare/v1.0.0-beta3...v1.0.0-beta4
[1.0.0-rc.1]: https://github.com/evpobr/sndfile2k/compare/v1.0.0-beta4...v1.0.0-rc.1

src/au.c
/*
** Copyright (C) 1999-2017 Erik de Castro Lopo <erikd@mega-nerd.com>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU Lesser General Public License as published by
** the Free Software Foundation; either version 2.1 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU Lesser General Public License for more details.
**
** You should have received a copy of the GNU Lesser General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include "config.h"

#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <ctype.h>

#include "sndfile2k/sndfile2k.h"
#include "sfendian.h"
#include "common.h"

/*
 * Macros to handle big/little endian issues.
 */

#define DOTSND_MARKER (MAKE_MARKER('.', 's', 'n', 'd'))
#define DNSDOT_MARKER (MAKE_MARKER('d', 'n', 's', '.'))

#define AU_DATA_OFFSET (24)

/*
 * Known AU file encoding types.
 */

enum
{
    /* 8-bit u-law samples */
    AU_ENCODING_ULAW_8 = 1,
    /* 8-bit linear samples */
    AU_ENCODING_PCM_8 = 2,
    /* 16-bit linear samples */
    AU_ENCODING_PCM_16 = 3,
    /* 24-bit linear samples */
    AU_ENCODING_PCM_24 = 4,
    /* 32-bit linear samples */
    AU_ENCODING_PCM_32 = 5,

    /* floating-point samples */
    AU_ENCODING_FLOAT = 6,
    /* double-precision float samples */
    AU_ENCODING_DOUBLE = 7,
    /* fragmented sampled data */
    AU_ENCODING_INDIRECT = 8,
    /* ? */
    AU_ENCODING_NESTED = 9,
    /* DSP program */
    AU_ENCODING_DSP_CORE = 10,
    /* 8-bit fixed-point samples */
    AU_ENCODING_DSP_DATA_8 = 11,
    /* 16-bit fixed-point samples */
    AU_ENCODING_DSP_DATA_16 = 12,
    /* 24-bit fixed-point samples */
    AU_ENCODING_DSP_DATA_24 = 13,
    /* 32-bit fixed-point samples */
    AU_ENCODING_DSP_DATA_32 = 14,

    /* non-audio display data */
    AU_ENCODING_DISPLAY = 16,
    /* ? */
    AU_ENCODING_MULAW_SQUELCH = 17,
    /* 16-bit linear with emphasis */
    AU_ENCODING_EMPHASIZED = 18,
    /* 16-bit linear with compression (NEXT) */
    AU_ENCODING_NEXT = 19,
    /* A combination of the two above */
    AU_ENCODING_COMPRESSED_EMPHASIZED = 20,
    /* Music Kit DSP commands */
    AU_ENCODING_DSP_COMMANDS = 21,
    /* ? */
    AU_ENCODING_DSP_COMMANDS_SAMPLES = 22,

    /* G721 32 kbs ADPCM - 4 bits per sample. */
    AU_ENCODING_ADPCM_G721_32 = 23,
    /* G722 64 kbs ADPCM */
    AU_ENCODING_ADPCM_G722 = 24,
    /* G723 24 kbs ADPCM - 3 bits per sample. */
    AU_ENCODING_ADPCM_G723_24 = 25,
    /* G723 40 kbs ADPCM - 5 bits per sample. */
    AU_ENCODING_ADPCM_G723_40 = 26,

    AU_ENCODING_ALAW_8 = 27
};

/*
 * Typedefs.
 */

typedef struct
{
    int dataoffset;
    int datasize;
    int encoding;
    int samplerate;
    int channels;
} AU_FMT;

/*
 * Private static functions.
 */

static int au_close(SF_PRIVATE *psf);

static int au_format_to_encoding(int format);

static int au_write_header(SF_PRIVATE *psf, int calc_length);
static int au_read_header(SF_PRIVATE *psf);

/*
 * Public function.
 */

int au_open(SF_PRIVATE *psf)
{
    int error = 0;

    if (psf->file.mode == SFM_READ || (psf->file.mode == SFM_RDWR && psf->filelength > 0))
    {
        if ((error = au_read_header(psf)))
            return error;
    };

    if ((SF_CONTAINER(psf->sf.format)) != SF_FORMAT_AU)
        return SFE_BAD_OPEN_FORMAT;

    int subformat = SF_CODEC(psf->sf.format);

    if (psf->file.mode == SFM_WRITE || psf->file.mode == SFM_RDWR)
    {
        psf->endian = SF_ENDIAN(psf->sf.format);
        if (CPU_IS_LITTLE_ENDIAN && psf->endian == SF_ENDIAN_CPU)
        {
            psf->endian = SF_ENDIAN_LITTLE;
        }
        else if (psf->endian != SF_ENDIAN_LITTLE)
        {
            psf->endian = SF_ENDIAN_BIG;
        }

        if (au_write_header(psf, SF_FALSE))
        {
            return psf->error;
        }

        psf->write_header = au_write_header;
    };

    psf->container_close = au_close;

    psf->blockwidth = psf->bytewidth * psf->sf.channels;

    switch (subformat)
    {
    case SF_FORMAT_ULAW: /* 8-bit Ulaw encoding. */
        ulaw_init(psf);
        break;

    case SF_FORMAT_PCM_S8: /* 8-bit linear PCM. */
        error = pcm_init(psf);
        break;

    case SF_FORMAT_PCM_16: /* 16-bit linear PCM. */
    case SF_FORMAT_PCM_24: /* 24-bit linear PCM */
    case SF_FORMAT_PCM_32: /* 32-bit linear PCM. */
        error = pcm_init(psf);
        break;

    case SF_FORMAT_ALAW: /* 8-bit Alaw encoding. */
        alaw_init(psf);
        break;

    case SF_FORMAT_FLOAT: /* 32-bit floats. */
        error = float32_init(psf);
        break;

    case SF_FORMAT_DOUBLE: /* 64-bit double precision floats. */
        error = double64_init(psf);
        break;

    case SF_FORMAT_G721_32:
        error = g72x_init(psf);
        psf->sf.seekable = SF_FALSE;
        break;

    case SF_FORMAT_G723_24:
        error = g72x_init(psf);
        psf->sf.seekable = SF_FALSE;
        break;

    case SF_FORMAT_G723_40:
        error = g72x_init(psf);
        psf->sf.seekable = SF_FALSE;
        break;

    default:
        break;
    };

    return error;
}

static int au_close(SF_PRIVATE *psf)
{
    if (psf->file.mode == SFM_WRITE || psf->file.mode == SFM_RDWR)
        au_write_header(psf, SF_TRUE);

    return 0;
}

static int au_write_header(SF_PRIVATE *psf, int calc_length)
{
    if (psf->pipeoffset > 0)
        return 0;

    sf_count_t current = psf_ftell(psf);

    if (calc_length)
    {
        psf->filelength = psf_get_filelen(psf);

        psf->datalength = psf->filelength - psf->dataoffset;
        if (psf->dataend)
        {
            psf->datalength -= psf->filelength - psf->dataend;
        }
    };

    int encoding = au_format_to_encoding(SF_CODEC(psf->sf.format));
    if (!encoding)
        return (psf->error = SFE_BAD_OPEN_FORMAT);

    /* Reset the current header length to zero. */
    psf->header.ptr[0] = 0;
    psf->header.indx = 0;

    /*
     * Only attempt to seek if we are not writng to a pipe. If we are
     * writing to a pipe we shouldn't be here anyway.
     */
    if (psf->is_pipe == SF_FALSE)
        psf_fseek(psf, 0, SEEK_SET);

    /*
     * AU format files allow a datalength value of -1 if the datalength
     * is not know at the time the header is written.
     * Also use this value of -1 if the datalength > 2 gigabytes.
     */
    
    int datalength;

    if (psf->datalength < 0 || psf->datalength > 0x7FFFFFFF)
        datalength = -1;
    else
        datalength = (int)(psf->datalength & 0x7FFFFFFF);

    if (psf->endian == SF_ENDIAN_BIG)
    {
        psf_binheader_writef(psf, "Em4", BHWm(DOTSND_MARKER), BHW4(AU_DATA_OFFSET));
        psf_binheader_writef(psf, "E4444", BHW4(datalength), BHW4(encoding),
                             BHW4(psf->sf.samplerate), BHW4(psf->sf.channels));
    }
    else if (psf->endian == SF_ENDIAN_LITTLE)
    {
        psf_binheader_writef(psf, "em4", BHWm(DNSDOT_MARKER), BHW4(AU_DATA_OFFSET));
        psf_binheader_writef(psf, "e4444", BHW4(datalength), BHW4(encoding),
                             BHW4(psf->sf.samplerate), BHW4(psf->sf.channels));
    }
    else
    {
        return (psf->error = SFE_BAD_OPEN_FORMAT);
    }

    /* Header construction complete so write it out. */
    psf_fwrite(psf->header.ptr, psf->header.indx, 1, psf);

    if (psf->error)
        return psf->error;

    psf->dataoffset = psf->header.indx;

    if (current > 0)
        psf_fseek(psf, current, SEEK_SET);

    return psf->error;
}

static int au_format_to_encoding(int format)
{
    switch (format)
    {
    case SF_FORMAT_PCM_S8:
        return AU_ENCODING_PCM_8;
    case SF_FORMAT_PCM_16:
        return AU_ENCODING_PCM_16;
    case SF_FORMAT_PCM_24:
        return AU_ENCODING_PCM_24;
    case SF_FORMAT_PCM_32:
        return AU_ENCODING_PCM_32;

    case SF_FORMAT_FLOAT:
        return AU_ENCODING_FLOAT;
    case SF_FORMAT_DOUBLE:
        return AU_ENCODING_DOUBLE;

    case SF_FORMAT_ULAW:
        return AU_ENCODING_ULAW_8;
    case SF_FORMAT_ALAW:
        return AU_ENCODING_ALAW_8;

    case SF_FORMAT_G721_32:
        return AU_ENCODING_ADPCM_G721_32;
    case SF_FORMAT_G723_24:
        return AU_ENCODING_ADPCM_G723_24;
    case SF_FORMAT_G723_40:
        return AU_ENCODING_ADPCM_G723_40;

    default:
        break;
    };
    return 0;
}

static int au_read_header(SF_PRIVATE *psf)
{
    AU_FMT au_fmt;
    int marker;

    memset(&au_fmt, 0, sizeof(au_fmt));
	psf_binheader_seekf(psf, 0, SF_SEEK_SET);
    psf_binheader_readf(psf, "m", &marker);
    psf_log_printf(psf, "%M\n", marker);

    if (marker == DOTSND_MARKER)
    {
        psf->endian = SF_ENDIAN_BIG;

        psf_binheader_readf(psf, "E44444", &(au_fmt.dataoffset), &(au_fmt.datasize),
                            &(au_fmt.encoding), &(au_fmt.samplerate), &(au_fmt.channels));
    }
    else if (marker == DNSDOT_MARKER)
    {
        psf->endian = SF_ENDIAN_LITTLE;
        psf_binheader_readf(psf, "e44444", &(au_fmt.dataoffset), &(au_fmt.datasize),
                            &(au_fmt.encoding), &(au_fmt.samplerate), &(au_fmt.channels));
    }
    else
    {
        return SFE_AU_NO_DOTSND;
    }

    psf_log_printf(psf, "  Data Offset : %d\n", au_fmt.dataoffset);

    if (psf->fileoffset > 0 && au_fmt.datasize == -1)
    {
        psf_log_printf(psf, "  Data Size   : -1\n");
        return SFE_AU_EMBED_BAD_LEN;
    };

    if (psf->fileoffset > 0)
    {
        psf->filelength = au_fmt.dataoffset + au_fmt.datasize;
        psf_log_printf(psf, "  Data Size   : %d\n", au_fmt.datasize);
    }
    else if (au_fmt.datasize == -1 || au_fmt.dataoffset + au_fmt.datasize == psf->filelength)
    {
        psf_log_printf(psf, "  Data Size   : %d\n", au_fmt.datasize);
    }
    else if (au_fmt.dataoffset + au_fmt.datasize < psf->filelength)
    {
        psf->filelength = au_fmt.dataoffset + au_fmt.datasize;
        psf_log_printf(psf, "  Data Size   : %d\n", au_fmt.datasize);
    }
    else
    {
        int dword = psf->filelength - au_fmt.dataoffset;
        psf_log_printf(psf, "  Data Size   : %d (should be %d)\n", au_fmt.datasize, dword);
        au_fmt.datasize = dword;
    };

    psf->dataoffset = au_fmt.dataoffset;
    psf->datalength = psf->filelength - psf->dataoffset;

	if (psf_ftell(psf) < psf->dataoffset)
		psf_binheader_seekf(psf, psf->dataoffset - psf_ftell(psf), SF_SEEK_CUR);

    psf->sf.samplerate = au_fmt.samplerate;
    psf->sf.channels = au_fmt.channels;

    /* Only fill in type major. */
    if (psf->endian == SF_ENDIAN_BIG)
        psf->sf.format = SF_FORMAT_AU;
    else if (psf->endian == SF_ENDIAN_LITTLE)
        psf->sf.format = SF_ENDIAN_LITTLE | SF_FORMAT_AU;

    psf_log_printf(psf, "  Encoding    : %d => ", au_fmt.encoding);

    psf->sf.format = SF_ENDIAN(psf->sf.format);

    switch (au_fmt.encoding)
    {
    case AU_ENCODING_ULAW_8:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_ULAW;
        psf->bytewidth = 1; /* Before decoding */
        psf_log_printf(psf, "8-bit ISDN u-law\n");
        break;

    case AU_ENCODING_PCM_8:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_PCM_S8;
        psf->bytewidth = 1;
        psf_log_printf(psf, "8-bit linear PCM\n");
        break;

    case AU_ENCODING_PCM_16:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_PCM_16;
        psf->bytewidth = 2;
        psf_log_printf(psf, "16-bit linear PCM\n");
        break;

    case AU_ENCODING_PCM_24:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_PCM_24;
        psf->bytewidth = 3;
        psf_log_printf(psf, "24-bit linear PCM\n");
        break;

    case AU_ENCODING_PCM_32:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_PCM_32;
        psf->bytewidth = 4;
        psf_log_printf(psf, "32-bit linear PCM\n");
        break;

    case AU_ENCODING_FLOAT:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_FLOAT;
        psf->bytewidth = 4;
        psf_log_printf(psf, "32-bit float\n");
        break;

    case AU_ENCODING_DOUBLE:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_DOUBLE;
        psf->bytewidth = 8;
        psf_log_printf(psf, "64-bit double precision float\n");
        break;

    case AU_ENCODING_ALAW_8:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_ALAW;
        psf->bytewidth = 1; /* Before decoding */
        psf_log_printf(psf, "8-bit ISDN A-law\n");
        break;

    case AU_ENCODING_ADPCM_G721_32:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_G721_32;
        psf->bytewidth = 0;
        psf_log_printf(psf, "G721 32kbs ADPCM\n");
        break;

    case AU_ENCODING_ADPCM_G723_24:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_G723_24;
        psf->bytewidth = 0;
        psf_log_printf(psf, "G723 24kbs ADPCM\n");
        break;

    case AU_ENCODING_ADPCM_G723_40:
        psf->sf.format |= SF_FORMAT_AU | SF_FORMAT_G723_40;
        psf->bytewidth = 0;
        psf_log_printf(psf, "G723 40kbs ADPCM\n");
        break;

    case AU_ENCODING_ADPCM_G722:
        psf_log_printf(psf, "G722 64 kbs ADPCM (unsupported)\n");
        break;

    case AU_ENCODING_NEXT:
        psf_log_printf(psf, "Weird NeXT encoding format (unsupported)\n");
        break;

    default:
        psf_log_printf(psf, "Unknown!!\n");
        break;
    };

    psf_log_printf(psf, "  Sample Rate : %d\n", au_fmt.samplerate);
    if (au_fmt.channels < 1)
    {
        psf_log_printf(psf, "  Channels    : %d  **** should be >= 1\n", au_fmt.channels);
        return SFE_CHANNEL_COUNT_ZERO;
    }
    else if (au_fmt.channels > SF_MAX_CHANNELS)
    {
        psf_log_printf(psf, "  Channels    : %d  **** should be <= %d\n", au_fmt.channels, SF_MAX_CHANNELS);
        return SFE_CHANNEL_COUNT;
    };

    psf_log_printf(psf, "  Channels    : %d\n", au_fmt.channels);

    psf->blockwidth = psf->sf.channels * psf->bytewidth;

    if (!psf->sf.frames && psf->blockwidth)
        psf->sf.frames = (psf->filelength - psf->dataoffset) / psf->blockwidth;

    return 0;
}
